File: apps/api/src/organizeg3_api/infrastructure/persistence/models/employee.py
"""SQLAlchemy ORM mapping for tenant employees."""

from __future__ import annotations

from datetime import date
import uuid

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Date,
    ForeignKeyConstraint,
    Index,
    Integer,
    String,
    UniqueConstraint,
    Uuid,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from organizeg3_api.infrastructure.database.base import (
    Base,
    TenantScopedMixin,
    TimestampMixin,
    UUIDPrimaryKeyMixin,
)


class EmployeeModel(
    UUIDPrimaryKeyMixin,
    TenantScopedMixin,
    TimestampMixin,
    Base,
):
    """Represent an employee belonging to one tenant."""

    __tablename__ = "employees"

    __table_args__ = (
        CheckConstraint(
            "TRIM(code) <> ''",
            name="code_not_blank",
        ),
        CheckConstraint(
            "TRIM(full_name) <> ''",
            name="full_name_not_blank",
        ),
        CheckConstraint(
            (
                "status IN ("
                "'ACTIVE', "
                "'ON_LEAVE', "
                "'INACTIVE', "
                "'TERMINATED'"
                ")"
            ),
            name="status_valid",
        ),
        UniqueConstraint(
            "tenant_id",
            "code",
            name="uq_employees_tenant_code",
        ),
        UniqueConstraint(
            "tenant_id",
            "document_number",
            name="uq_employees_tenant_document_number",
        ),
        UniqueConstraint(
            "tenant_id",
            "legacy_employee_id",
            name="uq_employees_tenant_legacy_employee_id",
        ),
        UniqueConstraint(
            "id",
            "tenant_id",
            name="uq_employees_id_tenant",
        ),
        ForeignKeyConstraint(
            [
                "branch_id",
                "tenant_id",
            ],
            [
                "branches.id",
                "branches.tenant_id",
            ],
            name="fk_employees_branch_tenant",
        ),
        Index(
            "ix_employees_tenant_active",
            "tenant_id",
            "is_active",
        ),
        Index(
            "ix_employees_tenant_status",
            "tenant_id",
            "status",
        ),
        Index(
            "ix_employees_tenant_branch",
            "tenant_id",
            "branch_id",
        ),
    )

    legacy_employee_id: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )

    branch_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid,
        nullable=True,
    )

    code: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    full_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )

    document_number: Mapped[str | None] = mapped_column(
        String(11),
        nullable=True,
    )

    email: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    phone: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
    )

    job_title: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    contract_type: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="ACTIVE",
        server_default=text("'ACTIVE'"),
    )

    birth_date: Mapped[date | None] = mapped_column(
        Date,
        nullable=True,
    )

    admission_date: Mapped[date | None] = mapped_column(
        Date,
        nullable=True,
    )

    termination_date: Mapped[date | None] = mapped_column(
        Date,
        nullable=True,
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
        server_default=text("true"),
    )

File: apps/pwa/src/features/employees/model/employee.ts
export type EmploymentStatus =
    | "ACTIVE"
    | "ON_LEAVE"
    | "INACTIVE"
    | "TERMINATED";

export interface Employee {
    readonly id: string;
    readonly tenant_id: string;
    readonly branch_id: string | null;
    readonly code: string;
    readonly full_name: string;
    readonly document_number: string | null;
    readonly email: string | null;
    readonly phone: string | null;
    readonly job_title: string | null;
    readonly contract_type: string | null;
    readonly status: EmploymentStatus;
    readonly birth_date: string | null;
    readonly admission_date: string | null;
    readonly termination_date: string | null;
    readonly is_active: boolean;
    readonly created_at: string | null;
    readonly updated_at: string | null;
}

export interface EmployeeCreateInput {
    readonly code: string;
    readonly full_name: string;
    readonly branch_id?: string | null;
    readonly document_number?: string | null;
    readonly email?: string | null;
    readonly phone?: string | null;
    readonly job_title?: string | null;
    readonly contract_type?: string | null;
    readonly birth_date?: string | null;
    readonly admission_date?: string | null;
}

export interface EmployeeUpdateInput {
    readonly code?: string;
    readonly full_name?: string;
    readonly branch_id?: string | null;
    readonly document_number?: string | null;
    readonly email?: string | null;
    readonly phone?: string | null;
    readonly job_title?: string | null;
    readonly contract_type?: string | null;
    readonly birth_date?: string | null;
    readonly admission_date?: string | null;
}

export interface EmployeeListFilters {
    readonly includeInactive?: boolean;
    readonly search?: string;
    readonly branchId?: string | null;
    readonly status?: EmploymentStatus | null;
    readonly limit?: number;
    readonly offset?: number;
}

export interface EmployeePage {
    readonly items: readonly Employee[];
    readonly hasPrevious: boolean;
    readonly hasNext: boolean;
    readonly offset: number;
    readonly pageSize: number;
}

export const DEFAULT_EMPLOYEE_PAGE_SIZE = 20;

export const EMPLOYMENT_STATUSES: readonly EmploymentStatus[] = [
    "ACTIVE",
    "ON_LEAVE",
    "INACTIVE",
    "TERMINATED",
];

export const EMPLOYMENT_STATUS_LABELS: Readonly<
    Record<EmploymentStatus, string>
> = {
    ACTIVE: "Ativo",
    ON_LEAVE: "Afastado",
    INACTIVE: "Inativo",
    TERMINATED: "Desligado",
};

export function getEmploymentStatusLabel(
    status: EmploymentStatus,
): string {
    return EMPLOYMENT_STATUS_LABELS[status];
}